prow/cache/cache.go

```diff
@@ -0,0 +1,203 @@
+// many concurrent threads all attempt to look up the same (missing) key/value
+// pair from the cache (see Alan Donovan and Brian Kernighan, "The Go
+// Programming Language" (Addison-Wesley, 2016), p. 277).
+
+// LRUCache is the actual concurrent non-blocking cache.
+type LRUCache struct {
+	*sync.Mutex
+	*simplelru.LRU
+}
+
+// Promise is a wrapper around cache value construction; it is used to
+// synchronize the to-be-cached value between threads that undergo a cache miss
+// and subsequent threads that attempt to look up the same cache entry.
+type Promise struct {
+	Result
+	resolve chan struct{}
+}
+
+// Result stores the result of executing an arbitrary ValConstructor function.
+type Result struct {
+	val interface{}
+	err error
+}
+
+// ValConstructor is used to construct a value. The assumption is that this
+// ValConstructor is expensive to compute, and that we need to memoize it via
+// the LRUCache. The raw values of a cache are only constructed after a cache
+// miss (and only the first cache miss). Using this type allows us to use any
+// arbitrary function whose resulting values needs to be memoized (saved in the
+// cache). This type also allows us to delay running the expensive computation
+// until we actually need it.
+type ValConstructor func() (interface{}, error)
+
+// NewLRUCache returns a new LRUCache with a given size (number of elements).
+func NewLRUCache(size int) (*LRUCache, error) {
+	cache, err := simplelru.NewLRU(size, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &LRUCache{&sync.Mutex{}, cache}, nil
+}
+
+// GetOrAdd tries to use a cache if it is available to get a Value. It is
+// assumed that Value is expensive to construct from scratch, which is the
+// reason why we try to use the cache in the first place. If we do end up
+// constructing a Value from scratch, then we store it into the cache with a
+// corresponding Key, so that we can look up the Value with just the Key in the
+// future.
+//
+// This cache is resistant to cache stampedes because it uses a duplicate
+// suppression strategy. This is also called request coalescing.
+func (lruCache *LRUCache) GetOrAdd(
+	key interface{},
+	valConstructor ValConstructor) (interface{}, error) {
+
+	// Cache lookup.
+	lruCache.Lock()
+	var promise *Promise
+	var ok bool
+	maybePromise, promisePending := lruCache.Get(key)
+
+	if promisePending {
+		// A promise exists, BUT the wrapped value inside it (p.val) might
+		// not be written to yet by the thread that is actually resolving the
+		// promise.
+		//
+		// For now we just unlock the overall lruCache itself so that it can
+		// service other GetOrAdd() calls to it.
+		lruCache.Unlock()
+
+		// If the type is not a promise type, there's no need to wait and we can
+		// just return immediately with an error.
+		promise, ok = maybePromise.(*Promise)
+		if !ok {
+			err := fmt.Errorf("Programmer error: expected cache entry type '*Promise', got '%T'", maybePromise)
+			logrus.Error(err)
```

Comment:
This log doesn't have context, suggest adding key to make this log more helpful
```suggestion
			logrus.WithField("key", key).Error(err)
```

Reply:
done